nntrainer/layers/conv2d_layer.cpp

```diff
@@ -244,85 +266,72 @@ static void im2col(const Tensor &in, const TensorDim &kdim,
   TensorDim bias_dim = TensorDim(1, filter_size, 1, 1);
 
-  if (weights.empty()) {
-    weights.reserve(2);
-    weights.emplace_back(dim, weight_initializer, weight_regularizer,
-                         weight_regularizer_constant, true, false,
-                         "Conv2d:filter");
-    weights.emplace_back(bias_dim, bias_initializer, WeightRegularizer::NONE,
-                         1.0f, true, false, "Conv2d:bias");
-    manager.trackWeights(weights);
-  } else {
-    weights[ConvParams::weight].reset(dim, weight_initializer,
-                                      weight_regularizer,
-                                      weight_regularizer_constant, true);
-    weights[ConvParams::bias].reset(bias_dim, bias_initializer,
-                                    WeightRegularizer::NONE, 1.0f, true);
-  }
+  wt_idx[ConvParams::weight] =
+    context.requestWeight(dim, weight_initializer, weight_regularizer,
+                          weight_regularizer_constant, "Conv2d:filter", true);
+  wt_idx[ConvParams::bias] =
+    context.requestWeight(bias_dim, bias_initializer, WeightRegularizer::NONE,
+                          1.0f, "Conv2d:bias", true);
 
   /// we don't have same padding for now but later, same padding don't apply
   /// when kernel size is even in current implementation (we need to handle
   /// assymetric padding)
 
   // this output_dim must be the same with dimension of hidden
+  TensorDim out_dim;
   out_dim.batch(in_dim.batch());
   out_dim.channel(filter_size);
   out_dim.height(
     (in_dim.height() - kernel_size[0] + 2 * padding[0]) / stride[0] + 1);
   out_dim.width((in_dim.width() - kernel_size[1] + 2 * padding[1]) / stride[1] +
                 1);
+  context.setOutputDimensions({out_dim});
 
   unsigned int eff_in_height = in_dim.height() + padding[0] * 2;
   unsigned int eff_in_width = in_dim.width() + padding[1] * 2;
 
   if (eff_in_height < kernel_size[0] || eff_in_width < kernel_size[1]) {
-    ml_loge("Failed to initialize: in size + padding is smaller than effective "
-            "kernel");
-    return ML_ERROR_INVALID_PARAMETER;
+    throw std::invalid_argument(
+      "Failed to initialize: in size + padding is smaller than effective "
+      "kernel");
   }
 
   unsigned int IM = std::numeric_limits<int>::max();
 
   if (eff_in_height - padding[0] - kernel_size[0] > IM ||
       eff_in_width - padding[1] - kernel_size[1] > IM) {
-    ml_loge("Failed to initialize: Calculated patch end is over int max");
-    return ML_ERROR_INVALID_PARAMETER;
+    throw std::invalid_argument(
+      "Failed to initialize: Calculated patch end is over int max");
   }
 
-  return status;
+  wt_idx[ConvParams::im2col_result] = context.requestTensor(
+    calcIm2ColOutputDim(in_dim, dim, padding, stride, {1, 1}), "Conv2d:im2col",
+    false, ITERATION_LIFESPAN);
+  wt_idx[ConvParams::col2im_result] = context.requestTensor(
+    calcCol2ImOutputDim(out_dim, dim), "Conv2d:col2im", false, BACKWARD_FUNC_LIFESPAN);
```

Comment:
If we ask with BACWARD_FUNC_LIFESPAN, then it created at backword process ? I mean, it looks like temperal variable only for the backwarding of this layer. so that, it does not need to alive all the way to the backwarding of network.

Reply:
> If we ask with BACWARD_FUNC_LIFESPAN, then it created at backword process ? I mean, it looks like temperal variable only for the backwarding of this layer. so that, it does not need to alive all the way to the backwarding of network. 

Yes, the variable will only be made available during the backward operation.

> And I wonder how manager decides it is temporal variable for this layer?

The manager will allocate the variable. However, the memory assigned to it will not be valid during the FORWARD. It is will be valid only during BACKWARD. RunLayerContext will ensure that this tensor is only requested in appropriate function.